qa/tasks/deepsea.py

```diff
@@ -363,18 +467,61 @@ def _introspect_roles(self, deepsea_ctx):
         for var in report_vars:
             self.log.info("{} == {}".format(var, deepsea_ctx[var]))
 
+    def _master_python_version(self, py_version):
+        """
+        Determine if a given python version is installed on the Salt Master
+        node.
+        """
+        python_binary = 'python{}'.format(py_version)
+        installed = True
+        try:
+            self.master_remote.run(args=[
+                'type',
+                python_binary,
+                run.Raw('>'),
+                '/dev/null',
+                run.Raw('2>&1'),
+                ])
+        except CommandFailedError:
+            installed = False
+        if installed:
+            self.master_remote.run(args=[
+                python_binary,
+                '--version'
+                ])
+        else:
+            self.log.info(
+                '{} not installed on master node'.format(python_binary)
+                )
+        return installed
+
+    def _maybe_apply_alternative_defaults(self):
+        global_yml = '/srv/pillar/ceph/stack/global.yml'
+        if self.alternative_defaults:
+            data = data = "\n".join([
+                yaml.dump(x, default_flow_style=False) for x in self.alternative_defaults
+                ])
+            if data:
+                # remote.sh does not return partial output when the command times out,
+                # so we hope the following command will not time out
+                self.master_remote.sh("sudo tee -=a {}".format(global_yml), input=data)
+        dump_file_that_might_not_exist(self.master_remote, global_yml)
+
     def _populate_deepsea_context(self):
         deepsea_ctx['roles'] = self.ctx.config['roles']
+        deepsea_ctx['alternative_defaults'] = self.config.get('alternative_defaults', {})
+        if not isinstance(deepsea_ctx['alternative_defaults'], dict):
+            raise ConfigError(self.err_prefix + "alternative_defaults must be a list")
```

Comment:
This is strange code, check if object is not a dict and complain that it must be a list.

Reply:
good catch